File: rails/test/services/checkout_service_test.rb
require "test_helper"

class CheckoutServiceTest < ActiveSupport::TestCase
  test "rejects invalid email" do
    error = assert_raises(CheckoutService::DemoError) do
      CheckoutService.parse("name" => "Akua", "email" => "bad", "phone" => "+233", "attempt_id" => "attempt_123")
    end
    assert_equal "validation_error", error.code
  end

  test "builds the shared order request" do
    checkout = CheckoutService::Checkout.new(name: "Akua", email: "akua@example.com", phone: "+233", attempt_id: "attempt_123")
    request = CheckoutService.order_params(checkout, "https://demo.example")
    assert_equal "demo-attempt_123", request.dig(:request_meta, :idempotency_key)
    assert_equal true, request[:finalize]
    assert_equal "https://demo.example/complete", request.dig(:checkout_settings, :redirect_url)
    assert_equal 5000, request.dig(:line_items, 0, :product, :price).value
  end
end

File: laravel/tests/Unit/CheckoutServiceTest.php
<?php

namespace Tests\Unit;

use App\Services\CheckoutService;
use Inttegro\Money\Currency;
use Inttegro\PriceParams;
use Inttegro\ProductType;
use PHPUnit\Framework\TestCase;

final class CheckoutServiceTest extends TestCase
{
    public function test_it_builds_the_shared_order_payload(): void
    {
        $product = [
            'type' => ProductType::Physical,
            'name' => 'Dawn Brew Set',
            'about' => 'A quiet ritual for slow mornings.',
            'reference' => 'DEMO-KORA-DAWN-BREW',
            'price' => new PriceParams(Currency::GHS, 5000),
        ];
        $payload = CheckoutService::orderPayload([
            'name' => 'Akua', 'email' => 'akua@example.com', 'phone' => '+233', 'attempt_id' => 'attempt_123',
        ], 'https://demo.example', $product);

        self::assertSame('demo-attempt_123', $payload['request_meta']['idempotency_key']);
        self::assertSame('KORA-ATTEMPT-123', $payload['number']);
        self::assertTrue($payload['finalize']);
        self::assertSame('https://demo.example/complete', $payload['checkout_settings']['redirect_url']);
        self::assertSame(ProductType::Physical, $payload['line_items'][0]['product']['type']);
        self::assertSame('Dawn Brew Set', $payload['line_items'][0]['product']['name']);
        self::assertSame(5000, $payload['line_items'][0]['product']['price']->value);
    }
}
